torch/csrc/autograd/VariableTypeUtils.h

```diff
@@ -134,88 +134,105 @@ template<typename... Args> inline variable_list flatten_tensor_args(Args&&... ar
-          view_func = [=](const at::Tensor& root_base) {
-            TORCH_CHECK(false, "This view is the output of a function that returns multiple views."
-                    "Such functions do not allow the output views to be modified inplace."
-                    "You should replace the inplace operation by an out-of-place one");
-            return root_base;
-          };
-        }
+        return make_variable_differentiable_view(std::move(tensor), ViewInfo(base, view_func),
+                                                 c10::nullopt, creation_meta, allow_tensor_metadata_change);
       }
-    } else if(diff_view_meta->has_view_fn()) {
-      // if current_view doesn't have a view_func but it's parent has one
-      auto prev_view_fn = diff_view_meta->view_fn();
-      auto size = tensor.sizes().vec();
-      auto stride = tensor.strides().vec();
-      auto storage_offset = tensor.storage_offset();
-      view_func = [=](const at::Tensor& root_base) {
-        auto temp = prev_view_fn(root_base);
-        return temp.as_strided(size, stride, storage_offset);
-      };
+    } else {
+      return make_variable_non_differentiable_view(base, std::move(tensor), allow_tensor_metadata_change);
     }
-    base_var = base_var._base();
   }
-  if (is_differentiable) {
-    return make_variable_differentiable_view(std::move(base_var), tensor, creation_meta, std::move(view_func));
+  // Create both the forward and backward info that are needed
+  c10::optional<ViewInfo> new_bw_info = c10::nullopt;
+  c10::optional<ViewInfo> new_fw_info = c10::nullopt;
+
+  if (is_bw_differentiable) {
+    if (base.is_view()) {
+      auto diff_view_meta = static_cast<DifferentiableViewMeta*>(torch::autograd::impl::get_autograd_meta(base));
+      auto base_bw_info = diff_view_meta->get_backward_view();
+      new_bw_info = base_bw_info.chain(base, tensor, view_func);
+    } else {
+      new_bw_info = ViewInfo(base, view_func);
+    }
   } else {
     TORCH_CHECK(creation_meta == CreationMeta::DEFAULT,
-                "Non-differentiable views must have creation_meta=CreationMeta::DEFAULT");
-    return make_variable_non_differentiable_view(std::move(base_var), tensor);
+                "Non-backward differentiable views must have creation_meta=CreationMeta::DEFAULT");
+  }
+
+  if (is_fw_differentiable) {
+    // Check if base is a forward differentiabble view
+    auto is_view = torch::autograd::impl::get_autograd_meta(base) && torch::autograd::impl::get_autograd_meta(base)->is_view_;
+    if (is_view && static_cast<DifferentiableViewMeta*>(torch::autograd::impl::get_autograd_meta(base))->has_fw_view()) {
+      auto diff_view_meta = static_cast<DifferentiableViewMeta*>(torch::autograd::impl::get_autograd_meta(base));
+      auto base_fw_info = diff_view_meta->get_forward_view();
+      new_fw_info = base_fw_info.chain(base, tensor, view_func);
+    } else {
+      new_fw_info = ViewInfo(base, view_func);
```

Comment:
You can eliminate the move assignment here by doing a trick with lambdas.

The problem is that you would like to define `new_fw_info` here (rather than above), but you need an if statement, and you don't want to stick it in a ternary (and even if you could, there are multiple statements.) So, what you do, is you make a small lambda that has the conditional:

```
ViewInfo new_fw_info = ([&]() {
  if () {
    ...
    return base_fw_info.chain(...);
  }
})();
```

and that will remove the move assignment; RVO should work.